rfcs/2021-09-14-3910-schema-support.md

```diff
@@ -0,0 +1,472 @@
+RFC3339, while another uses Unix time.
+
+This last example shows that there are details that we still need to flesh out.
+For example, do we allow loss of data when converting events between schema's,
+as long as this is documented (e.g. RFC3339 supports a wider range of dates than
+Unix time does).
+
+To use this automated conversion, all the operator needs to do, is connect the
+relevant source (optionally through a pipeline of transforms) to the final sink.
+Vector can detect at boot which schema the sink is going to receive based on the
+pipelines involved, check whether the sink type has a schema attached to it, and
+whether that schema can be converted to, from the received event schemas. If it
+can't, Vector refuses to boot, and explains the reasoning for not booting.
+
+#### Remap-based Conversion
+
+If automated schema conversion is not available, or unwanted, operators can
+leverage the `remap` transform to do a manual schema conversion.
+
+There are multiple ways this transform ties into the proposed schema support.
+
+##### Type Checking Integration
+
+First, when an event received by the transform has a schema attached to it, the
+VRL compiler uses this schema during type checking. Before this RFC, all event
+fields started as "any" type (e.g. unknown). Once this RFC is implemented,
+schema type information can be used to inform the VRL compiler. This means that
+operators no longer have to do the type enforcement themselves, meaning
+a program such as this:
+
+```coffee
+.http_code = to_int(.http_code) ?? 0
+if (.http_code >= 400 && .http_code < 500) {
+  # ...
+}
+```
+
+Can now be simplified to:
+
+```coffee
+if (.http_code >= 400 && .http_code < 500) {
+  # ...
+}
+```
+
+Because the compiler infers from the schema that the `http_code` field must
+always be an integer.
+
+It goes without saying that allowing this to exists means that the VRL compiler
+is entirely dependent on Vector providing the **correct** schema details for any
+given event. If it does not, and the received event is shaped differently from
+what the VRL compiler expected based on the schema definition it received,
+undefined runtime errors will occur when running the VRL program against the
+event.
+
+##### Output Schema Validation
+
+Additionally, if the operator specified the required output schema of the
+transform, Vector uses VRL's type checking to ensure that whatever the VRL
+program produces once it runs to completion, matches the expected output schema.
+
+If the output does not match the expected schema, Vector won't boot. This
+ensures that operators can update VRL programs while knowing that Vector will
+continue to validate the final output, preventing them from introducing a bug in
+the program that invalidates the schema adherence.
+
+```toml
+[transforms.remap]
+  type = "remap"
+  inputs = []
+
+  # Ensure any event produced by this transform adheres to my "custom" schema.
+  schema.out.type = "custom"
+  schema.out.file = "./my_schema.json"
```

Comment:
I think we could allow inline definition too. This might be related to a general user desire to "include" other files in a Vector config.